crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,214 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    fs,
+    fs::File,
+    hash::{Hash, Hasher},
+    io::{BufReader, BufWriter},
+    path::PathBuf,
+};
+
+use matrix_sdk::oidc::types::registration::VerifiedClientMetadata;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, thiserror::Error)]
+pub enum OidcRegistrationsError {
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+}
+
+/// The data needed to restore an OpenID Connect session.
+#[derive(Debug)]
+pub struct OidcRegistrations {
+    /// The path of the file where the registrations are stored.
+    file_path: PathBuf,
+    /// The hash for the metadata used to register the client.
+    /// This is used to check if the client needs to be re-registered.
+    metadata_hash: u64,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RegistrationData {
+    /// The hash for the metadata used to register the client.
+    metadata_hash: u64,
+    /// All of the registrations this client has made as a HashMap of issuer URL
+    /// (as a string) to client ID (as a string).
+    dynamic_registrations: HashMap<String, String>,
```

Comment:
For this field and the static one in `OidcRegistrations`, could we use wrapper types to indicate what the string represent? Maybe we could use an Url type for the key (ISTR we're using some in the code base), and for the value something like:

```rust
struct ClientId(String);
```

Reply:
Ah yeah, I think in strings instead of URLs in the FFI because of uniffi. That definitely makes sense.